### ISSUE TITLE
How to use an LSTM Layer in CNN?

### ISSUE THREAD

**Issue by II-Matto** [apache/incubator-mxnet#8908]
I would like to use an LSTM layer in CNN. Specifically, the network will first have a few convolutional layers, producing output `y1` from input `x`. Then `y1` is interpreted as a sequence and fed into an LSTM, producing output `y2` at the last step. Finally `y1` and `y2` is fed together into one or more layers, which can be convolutional, fully-connected or any other reasonable layers, to produce the final output `z`. The whole network is supposed to be trained jointly in an end-to-end manner.
```
x ---(CONV)---> y1 ---(LSTM)---> y2
[y1, y2] ---(CONV/FC/...)---> z
```
How can I create the symbol of such a network with MXNet?

--------

I can conveniently achieve this in Caffe. The only thing that I need to do is adding an LSTM layer in prototxt just like adding any other layers.

(BTW, the size of inputs of the network will be determined on the fly, and module binding is performed each time the input size changes.)

Currently I am considering to use the [RNN Cell Symbol API](https://mxnet.incubator.apache.org/api/python/symbol/rnn.html). But it seems that problems exist. For example,

1. I won't be able to `reset the cell when unrolling/stepping for a new sequence by calling lstm_cell.reset()`, since the `LSTMCell` is only used when defining the symbol of the network.
2. Loading or saving RNN models are supposed to be done by calling RNN-specific functions such as `mx.rnn.do_rnn_checkpoint()`, which seem to be incompatible with non-RNN layers/operators.

If using the RNN Cell Symbol API is a feasible solution, can someone provide an accessible example?

**Comment by eric-haibin-lin** [comment]
@szha can we do that in Gluon? 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by kevinthesun** [comment]
You can try Gluon API, which is more flexible for such kind of complicate network.
A simple sketch example for your network:
```python
class CNNLSTM(HybridBlock):
    def __init__(self, ...):
        self.cnn_block = ...
        self.lstm_block = mxnet.gluon.rnn.LSTM(...)
        self.fc_block = ...

    def hybrid_forward(F, x):
        y1 = self.cnn_block(x)
        y2 = self.lstm_block(y1)
        y3 = F.concat(y1, y2)
        z = self.fc_block(y3)
        return z
```

**Comment by szha** [comment]
@II-Matto if your computation graph doesn't change, which means `x` and `y1` always have the same shapes, then you only need to unroll the graph once. If your computation graph does change based on sequence length, then the computation graph is a function based on input length/shape, in which case you would need to define a `sym_gen` function as you normally would when using mx.rnn. For basic usage, you can find an example at https://github.com/apache/incubator-mxnet/blob/master/example/rnn/lstm_bucketing.py#L88-L103.

Regarding having CNN, there isn't anything special about adding a conv layer before the RNN layer. For example, you can use the layout (batch size, time steps, input dimensions) (i.e. `NTC`) for input data. After convolution layer, all you need to do is to make sure that the output is reshaped according to rnn API's expectation before using it to unroll the rnn cell.

The same functionality as mx.rnn can be found in `mx.gluon.rnn`. If you don't need special modification other than dropout, the rnn layers in gluon is probably the easiest to use. You can find an example at the beginning of its doc at https://mxnet.incubator.apache.org/versions/master/api/python/gluon/rnn.html?highlight=gluon.rnn#recurrent-layers. This would be the closest to what caffe LSTM layer offers.

**Comment by II-Matto** [comment]
@eric-haibin-lin @kevinthesun @szha Thanks for your prompt replies!

For a temporary solution, I borrowed the code from the [Warp CTC example](https://github.com/apache/incubator-mxnet/tree/master/example/warpctc), which implements [LSTM](https://github.com/apache/incubator-mxnet/blob/master/example/warpctc/lstm.py) with general fully-connected layers. I used the `lstm()` and `lstm_unroll()` functions and replaced the `wordvec` with my symbol of convolutional features. The model can be trained normally, though some of the losses haven't decreased to an expected level up to now. I will give more feedback after the training finishes.

-------

The solution provided by @kevinthesun is nice. But I'm not sure if it can right fit in my current code.

I'm using the [R-CNN example](https://github.com/apache/incubator-mxnet/tree/master/example/rcnn) for object detection. The example provides symbol definition files of VGG-16 and ResNet, and they are called in the training script to generate the corresponding symbol, which is then initialized and passed to a module for network training. From my understanding, to use a different network, what I need to do is to provide the corresponding symbol definition file. This will leave other code untouched.

@kevinthesun, if I would like to use your solution, then I will need to write a problem solver (or module) to handle the forward and backward process on my own. Am I right?

-------

@szha, your first solution uses the [RNN Cell Symbol API](https://mxnet.incubator.apache.org/api/python/symbol/rnn.html) as I requested. It is well explained and looks consistent with other symbol definitions. In this case, should I call the RNN-specific functions for model saving and loading or a normal `do_checkpoint()` is enough? And is it OK that the RNN cell variable together with the `sym_gen()` function are defined in a separate file (i.e. the file creating the module will import the `sym_gen()` from another file)?

Your second solution suggests to use the `mx.gluon.rnn` from Gluon API. This seems to be a more favorable solution since both regular neural network layers and recurrent ones can be used with `mx.gluon.nn.Sequential()`. My question is: can the `mx.gluon.nn.Sequential` object be used as an MXNet symbol? More specifically, can a Gluon `Sequential` object be passed to a function that originally accepts an MXNet symbol? This is important for me since the [R-CNN example](https://github.com/apache/incubator-mxnet/tree/master/example/rcnn) that I'm using is not implemented with Gluon.

**Comment by szha** [comment]
`do_rnn_checkpoint` is required when mx.rnn cells are used, and it would do checkpointing for both rnn cells and regular layer weights.

Sequential object itself can't be used to generate symbol. With Gluon, you won't have to use symbol. Gluon is generally a more favorable API to use if you are developing new models. When developing new models, you can run the model in imperative mode to immediately examine the outputs, making it very easy to debug. When development finishes, the gluon interface allows caching and reusing the computation graph parts that won't change to achieve close-to-optimal speed.

Assuming that you want to perform object detection on videos, you may want to check out what @zhreshold has been working on with object detection with gluon API in #8639, in which SSD and YOLOv2 are completed, and faster R-CNN and mask R-CNN are on his roadmap.

**Comment by kevinthesun** [comment]
@II-Matto In gluon, you just need to define forward pass. Backward can be completed by autograd.

**Comment by II-Matto** [comment]
@szha @kevinthesun Many thanks for your help. I will check out if it is possible to switch to use Gluon in my project at present.

BTW, the PR https://github.com/apache/incubator-mxnet/pull/8639 is very exciting, and I can't wait to see its completion!
